### DISCUSSION TITLE
Not a developer question but regular github user, keeping tools updated?

### DISCUSSION THREAD

**Discussion by mathurin68** [community/community#23337] [New to GitHub]
<p>I’m not a developer(yet) BUT I do grab a lot of security tools and code that people post on github(awesome).  </p>
<p>If I download a tool with -</p>
<p>git clone ‘path to tool.git’</p>
<p>and then I work with a file in that folder, say a json template, a text file or something like that. </p>
<p>Then, a couple months later, there’s an awesome improvement to the tool which I want so I run again - </p>
<p>git clone ‘path to tool.git’    </p>
<p>BUT, now I get the ‘merge’ problems.  </p>
<p>I understand git is protecting my changes but what’s the best way to keep this updated?  </p>
<p>I’ve tried</p>
<p>‘git commit -m “my message”’</p>
<p>git stash</p>
<p>git stash pop</p>
<p>and almost everytime there’s problems(I’m sure it’s me). </p>
<p>The easiest thing so far has just been to rename the first folder, git clone the latest, and then copy over my modified files.  </p>
<p>Is there an easier ‘git’ way to do this?  </p>
<p>Thank you and stay safe!</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>In principle merge conflicts happen when both you and the upstream repository have changed the same files, in ways that can’t be automatically resolved. You will have to take some kind of manual steps to resolve that. In priciple your current approach of copying your files over a fresh clone does that, too: Keep your changes, discard any remote changes to those files.</p>
<p>The git way to handle this would be to use branches: Create a branch for your changes and commit them to that branch. That way you get to manage your changes with git. When upstream has changes you want you then have two options:</p>
<ol>
<li>
<a href="https://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging" rel="nofollow noopener">Merge the upstream branch</a> (e.g. master) into your branch.</li>
<li>
<a href="https://git-scm.com/book/en/v2/Git-Branching-Rebasing" rel="nofollow noopener">Rebase your branch</a> onto the updated upstream branch.</li>
</ol>
<p>In both cases you will have to resolve any conflicts, the section of the <a href="https://git-scm.com/book/en/v2" rel="nofollow noopener">Git Book</a> I linked regarding merging also explains how to do that. I’d recommend reading the first three chapters to get a little more context.</p>

**Comment by mathurin68** [comment]
<p>Got it… thanks <a class="mention" href="/airtower-luna">@airtower-luna</a> !!</p>
